community/values/src/main/java/org/neo4j/values/storable/UTF8StringValue.java

```diff
@@ -0,0 +1,445 @@
+    public boolean equals( Value value )
+    {
+        if ( value instanceof UTF8StringValue )
+        {
+            return Arrays.equals( bytes, ((org.neo4j.values.storable.UTF8StringValue) value).bytes );
+        }
+        else
+        {
+            return super.equals( value );
+        }
+    }
+
+    @Override
+    String value()
+    {
+        String s = value;
+        if ( s == null )
+        {
+            synchronized ( this )
+            {
+                s = value;
+                if ( s == null )
+                {
+                    value = s = new String( bytes, offset, byteLength, StandardCharsets.UTF_8 );
+                }
+            }
+        }
+        return s;
+    }
+
+    @Override
+    public int length()
+    {
+        int count = 0, i = offset, len = offset + byteLength;
+        while ( i < len )
+        {
+            byte b = bytes[i];
+            //If high bit is zero (equivalent to the byte being positive in two's complement)
+            //we are dealing with an ascii value and use a single byte for storing the value.
+            if ( b >= 0 )
+            {
+                i++;
+            }
+
+            //The number of high bits tells us how many bytes we use to store the value
+            //e.g. 110xxxx -> need two bytes, 1110xxxx -> need three bytes, 11110xxx -> needs
+            //four bytes
+            while ( b < 0 )
+            {
+                i++;
+                b = (byte) (b << 1);
+            }
+            count++;
+        }
+        return count;
+    }
+
+    @Override
+    public int computeHash()
+    {
+        byte[] values = bytes;
+
+        if ( values.length == 0 || byteLength == 0 )
+        {
+            return 0;
+        }
+
+        int hash = 1, i = offset, len = offset + byteLength;
+        while ( i < len )
+        {
+            byte b = values[i];
+            //If high bit is zero (equivalent to the byte being positive in two's complement)
+            //we are dealing with an ascii value and use a single byte for storing the value.
+            if ( b >= 0 )
+            {
+                hash = 31 * hash + b;
+                i++;
+                continue;
+            }
+
+            //We can now have one of three situations.
+            //Byte1    Byte2    Byte3    Byte4
+            //110xxxxx 10xxxxxx
+            //1110xxxx 10xxxxxx 10xxxxxx
+            //11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+            //Figure out how many bytes we need by reading the number of leading bytes
+            int bytesNeeded = 0;
+            while ( b < 0 )
+            {
+                bytesNeeded++;
+                b = (byte) (b << 1);
+            }
+            int codePoint = codePoint( b, i, bytesNeeded );
+            i += bytesNeeded;
+
+            hash = 31 * hash + codePoint;
+        }
+
+        return hash;
+    }
+
+    @Override
+    public TextValue substring( int start, int length )
+    {
+        if ( start < 0 || length < 0)
```

Comment:
The style god demands that you sacrifice a whitespace and add it before the closing paren on this line.